pkg/provision/cleanup.go

```diff
@@ -0,0 +1,230 @@
+package provision
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zbus"
+	"github.com/threefoldtech/zos/pkg"
+	"github.com/threefoldtech/zos/pkg/storage/filesystem"
+	"github.com/threefoldtech/zos/pkg/stubs"
+	"github.com/threefoldtech/zos/pkg/zdb"
+	"golang.org/x/net/context"
+)
+
+func execute(cmd string, args ...string) ([]byte, error) {
+	exe := exec.Command(cmd, args...)
+	output, err := exe.CombinedOutput()
+	if err != nil {
+		return nil, errors.Wrapf(err, "execute '%v %+v': %v", cmd, args, string(output))
+	}
+
+	return output, nil
+}
+
+func lines(input []byte) []string {
+	input = bytes.TrimSpace(input)
+	if len(input) == 0 {
+		return nil
+	}
+
+	return strings.Split(string(input), "\n")
+}
+
+func must(input []byte, err error) []byte {
+	if err != nil {
+		panic(err)
+	}
+
+	return input
+}
+
+type container struct {
+	ID   string `json:"ID"`
+	Spec struct {
+		Root struct {
+			Path string `json:"path"`
+		} `json:"root"`
+		Mounts []struct {
+			Source string `json:"source"`
+		} `json:"mounts"`
+	} `json:"Spec"`
+}
+
+// we declare this method as a variable so we can
+// mock it in testing.
+var zdbConnection = func(id string) zdb.Client {
+	socket := fmt.Sprintf("unix://%s/zdb.sock", socketDir(id))
+	return zdb.New(socket)
+}
+
+func socketDir(containerID string) string {
+	return fmt.Sprintf("/var/run/zdb_%s", containerID)
+}
+
+// CleanupResources cleans up unused resources
+func CleanupResources() error {
+	toSave := make(map[string]struct{})
+	toDelete := make(map[string]struct{})
+
+	for _, ns := range lines(must(execute("ctr", "namespace", "ls", "-q"))) {
+		log.Info().Msgf("Checking namespace %s", ns)
+		for _, cnt := range lines(must(execute("ctr", "-n", ns, "container", "ls", "-q"))) {
+			bytes := must(execute("ctr", "-n", ns, "container", "info", cnt))
+			var container container
+			if err := json.Unmarshal(bytes, &container); err != nil {
+				panic(err)
+			}
+
+			zdbCl := zdbConnection(container.ID)
+			defer zdbCl.Close()
+			if err := zdbCl.Connect(); err != nil {
+				return errors.Wrapf(err, "failed to connect to 0-db: %s", container.ID)
+			}
+
+			ns, err := zdbCl.Namespaces()
+			if err != nil {
+				return errors.Wrap(err, "failed to retrieve zdb namespaces")
```

Comment:
General notice about early returns here. This is after all a clean up method, so it should do its best to bring the system to a consistent and clean state, right? 
It means it should be able to "skip" some non fatal issues. For example if a zdb instance is not responding this shouldn't stop the system from checking other instances.

And next time this clean up will run it will recheck this lingering one